allennlp/training/optimizers.py

```diff
@@ -196,6 +205,159 @@ def default(model_parameters: List) -> "Optimizer":
         return Optimizer.from_params(model_parameters=model_parameters, params=Params({}))
 
 
+@Optimizer.register("regex")
+class RegexOptimizer(Optimizer):
+    """
+    A `RegexOptimizer` creates a dictionary of `Optimizer`s keyed on some 'name'.
+    Each Optimizer contains its own set of parameters which are obtained using
+    regex matches for certain model parameters.
+    
+    This optimizer works by taking in a parameter `optimizers` which contains a list of `Optimizers`
+    with their keyword arguments, and a parameter `parameter_groups`, which contains regexes and their
+    corresponding optimizer and optional non-default optimizer options for this group. 
+    The regexes in the parameter groups are assigned to their optimizer based on the 'name' argument
+    where the 'name' value should be the same for the optimizer and parameter group.
+    You should specify a default optimizer with 'name': 'default' which will be used for all 
+    parameters which didn't obtain a regex match or when your parameter group doesn't contain a 'name' 
+    parameter.
+
+    # Parameters
+
+    optimizers: `List[Dict[str, Any]]`
+        A list of optimizers to use. Each entry in the list is a dictionary of keyword arguments. A 'name' 
+        keyword argument should be given which will serve as the key to match the optimizer with a 
+        specific parameter group. You should also supply an entry for the default parameter group, 
+        e.g. 'name': 'default'.
+
+    parameter_groups:  `List[Tuple[List[str], Dict[str, Any]]`, optional (default = `None`)
+        See the docstring of `make_parameter_groups` for what this parameter should look like. It
+        should follow the same format as there, except an additional 'name' argument should be provided
+        to match this group to its own optimizer. Optimizer options can also be set for this group which
+        will override the default options.
+    """
+
+    def __init__(
+        self,
+        model_parameters: List[Tuple[str, torch.nn.Parameter]],
+        optimizers: List[Dict[str, Any]],
+        parameter_groups: List[Tuple[List[str], Dict[str, Any]]],
+    ):
+        self.model_parameters = model_parameters
+        self.optimizers = optimizers
+        self.parameter_groups = parameter_groups
+
+        self._grouped_optimizers: Dict[str, Optimizer] = {}
+
+        optimizer_names = [optimizer["name"] for optimizer in self.optimizers]
+        assert "default" in optimizer_names, (
+           "No optimizer was provided for the 'default' group."
+           " Please provide a 'name': 'default' parameter for one of the `optimizers`."
+           )
+
+        parameter_groups = make_parameter_groups(self.model_parameters, self.parameter_groups)
+        # Manually set optimizer kwargs for the default group, so certain schedulers are able
+        # to know the options for this group.
+        if len(list(parameter_groups[-1].keys())) == 1:
```

Comment:
By default, `make_parameter_groups` does not assign any optimizer options to the last (default) group. These options are set when the optimizer is initialized. But as this class is initialized with these `parameter_groups` I added these manually for the default group so they were visible to a `learning_rate_scheduler`.